meteor/server/publications/partsUI/publication.ts

```diff
@@ -0,0 +1,284 @@
+		const partPosition = findPartPosition(part, segmentRanks, rundownRanks)
+		const isLoopingOverriden =
+			isLoopDefined &&
+			playlist.quickLoop?.forceAutoNext !== ForceQuickLoopAutoNext.DISABLED &&
+			compareMarkerPositions(quickLoopStartPosition, partPosition) >= 0 &&
+			compareMarkerPositions(partPosition, quickLoopEndPosition) >= 0
+
+		if (isLoopingOverriden && (part.expectedDuration ?? 0) <= 0) {
+			if (playlist.quickLoop?.forceAutoNext === ForceQuickLoopAutoNext.ENABLED_FORCING_MIN_DURATION) {
+				part.expectedDuration = studio.settings.fallbackPartDuration ?? DEFAULT_FALLBACK_PART_DURATION
+				part.expectedDurationWithPreroll =
+					studio.settings.fallbackPartDuration ?? DEFAULT_FALLBACK_PART_DURATION
+			} else if (playlist.quickLoop?.forceAutoNext === ForceQuickLoopAutoNext.ENABLED_WHEN_VALID_DURATION) {
+				part.invalid = true
+				part.invalidReason = {
+					message: generateTranslation('Part duration is 0.'),
+				}
+			}
+		}
+		part.autoNext = part.autoNext || (isLoopingOverriden && (part.expectedDuration ?? 0) > 0)
+	}
+
+	updateProps?.invalidatePartIds?.forEach((partId) => {
+		collection.remove(partId) // if it still exists, it will be replaced in the next step
+	})
+
+	const invalidatedSegmentsSet = new Set(updateProps?.invalidateSegmentIds ?? [])
+	const invalidatedPartsSet = new Set(updateProps?.invalidatePartIds ?? [])
+
+	state.contentCache.Parts.find({}).forEach((part) => {
+		if (
+			updateProps?.invalidateQuickLoop ||
+			invalidatedSegmentsSet.has(part.segmentId) ||
+			invalidatedPartsSet.has(part._id)
+		) {
+			modifyPartForQuickLoop(part)
+			collection.replace(part)
+		}
+	})
+}
+
+function findMarkerPosition(
+	marker: QuickLoopMarker,
+	fallback: number,
+	contentCache: ReadonlyObjectDeep<ContentCache>,
+	rundownRanks: Record<string, number>
+): MarkerPosition {
+	const part = marker.type === QuickLoopMarkerType.PART ? contentCache.Parts.findOne(marker.id) : undefined
+	const partRank = part?._rank ?? fallback
+
+	const segmentId = marker.type === QuickLoopMarkerType.SEGMENT ? marker.id : part?.segmentId
+	const segment = segmentId && contentCache.Segments.findOne(segmentId)
+	const segmentRank = segment?._rank ?? fallback
+
+	const rundownId = marker.type === QuickLoopMarkerType.RUNDOWN ? marker.id : segment?.rundownId
+	let rundownRank = rundownId ? rundownRanks[unprotectString(rundownId)] : fallback
+
+	if (marker.type === QuickLoopMarkerType.PLAYLIST) rundownRank = fallback
+
+	return {
+		rundownRank: rundownRank,
+		segmentRank: segmentRank,
+		partRank: partRank,
+	}
+}
+
+function findPartPosition(
+	part: DBPart,
+	segmentRanks: Record<string, number>,
+	rundownRanks: Record<string, number>
+): MarkerPosition {
+	return {
+		rundownRank: rundownRanks[part.rundownId as unknown as string] ?? 0,
+		segmentRank: segmentRanks[part.segmentId as unknown as string] ?? 0,
+		partRank: part._rank,
+	}
+}
+
+function stringsToIndexLookup(strings: string[]): Record<string, number> {
```

Comment:
It might be nicer if this returned a Map, then it could preserve the `RundownId` or whatever `ProtectedString` typings